playbooks/files/backup/backup.sh

```diff
@@ -0,0 +1,150 @@
+    then
+      kpartx -l /dev/$volume-snap | while read loop_device _
+      do
+        findmnt /dev/mapper/$loop_device -o TARGET | while read mount_point
+        do
+	  umount /dev/mapper/$loop_device
+	  rmdir $mount_point
+	done
+	test -f $(get_backup_location $loop_device) &&
+          rm -f $(get_backup_location $loop_device)
+      done
+      kpartx -d /dev/$volume-snap
+      rm -f $BACKUP_FROM_DIR/$volume.mbr
+      rm -f $BACKUP_FROM_DIR/$volume.sfdisk
+    else
+      mountpoint -q $BACKUP_FROM_DIR/$volume &&
+        umount $BACKUP_FROM_DIR/$volume &&
+	  rmdir $BACKUP_FROM_DIR/$volume
+      test -f $BACKUP_FROM_DIR/$volume &&
+        rm $BACKUP_FROM_DIR/$volume
+    fi
+    /usr/sbin/lvremove --yes $volume-snap
+  done
+}
+
+test ! "${RESTIC_REPOSITORY+x}" -a ! "${RESTIC_REPOSITORY_FILE+x}" &&
+  echo "ERROR: Environment variable RESTIC_REPOSITORY or RESTIC_REPOSITORY_FILE should be defined" > /dev/stderr &&
+    exit 1
+
+test ! "${RESTIC_PASSWORD+x}" -a ! "${RESTIC_PASSWORD_COMMAND+x}" -a ! "${RESTIC_PASSWORD_FILE+x}" &&
+  echo "ERROR: Environment variable RESTIC_PASSWORD, RESTIC_PASSWORD_COMMAND or RESIC_PASSWORD_FILE should be defined" > /dev/stderr &&
+    exit 1
+
+trap teardown HUP INT QUIT ABRT ALRM TERM KILL
+
+teardown
+
+# Backup MBR and partition tables of harddrives
+mkdir -p $BACKUP_FROM_DIR
+lsblk -l -o TYPE,NAME | grep "^disk " | while read _ harddrive
+do
+  dd if=/dev/$harddrive of=$BACKUP_FROM_DIR/$harddrive.mbr bs=512 count=1
+  sfdisk -d /dev/$harddrive > $BACKUP_FROM_DIR/$harddrive.sfdisk
+done
+# Backup boot and EFI
+mount -m $(findmnt -n -o SOURCE /boot) $BACKUP_FROM_DIR/boot
+mount -m $(findmnt -n -o SOURCE /boot/efi) $BACKUP_FROM_DIR/boot_efi
+# Backup filesystems
+get_volumes | while read volume
+do
+  /usr/sbin/lvcreate --snapshot -L 1g -n $volume-snap $volume
+  # The created snapshot can contain a disk with partitions or a filesystem directly.
+  # If it is a disk with partitions a MBR and a partition table dump is performed.
+  # Moreover, each partition in the partition table is mounted if contains a filesystem
+  # or dumped if it does not contain a filesystem (BIOS boot).
+  # Otherwise, if the created snapshot does not contain a disk with partitions, it is
+  # assumed that contains a filesystem directly and it is mounted. If mount fails a dump
+  # is created.
+  if is_partition_table /dev/$volume-snap
+  then
+    dd if=/dev/$volume-snap of=$BACKUP_FROM_DIR/$volume.mbr bs=512 count=1
+    sfdisk -d /dev/$volume-snap > $BACKUP_FROM_DIR/$volume.sfdisk
+    kpartx -a /dev/$volume-snap
+    kpartx -l /dev/$volume-snap | while read loop_device _
+    do
+      if is_filesystem /dev/mapper/$loop_device
+      then
+	mount -m -o ro /dev/mapper/$loop_device $(get_backup_location $loop_device)
+      else
+        dd if=/dev/mapper/$loop_device of=$(get_backup_location $loop_device)
+      fi
+    done
+  else
+    if ! mount -m -o ro /dev/$volume-snap $BACKUP_FROM_DIR/$volume
```

Comment:
why not use `is_filesystem` here?

Reply:
We can use it and it will work now, but the file system detection should be improved because if there is a swap partition, it is detected incorrectly. I'm going to document it.

Reply:
Done via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/1d0df128f8efb47628027e8ba61d39b610af117a

Reply:
I improved filesystem detection via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/00e5a224b88e3bc9469c6ac794c0fda7392fdc1b